stl/inc/vector

```diff
@@ -440,6 +440,183 @@ constexpr _Ty* _Unfancy_maybe_null(_Ty* _Ptr) noexcept { // do nothing for plain
     return _Ptr;
 }
 
+#ifdef __SANITIZE_ADDRESS__
+extern "C" void __sanitizer_annotate_contiguous_container(const void*, const void*, const void*, const void*);
+
+template <class _Vec, class = void>
+constexpr bool _Has_minimum_allocation_alignment = disjunction_v<bool_constant<alignof(typename _Vec::value_type) >= 8>,
+    is_same<typename _Vec::allocator_type, allocator<typename _Vec::value_type>>>;
+
+template <class _Vec>
+constexpr bool
+    _Has_minimum_allocation_alignment<_Vec, void_t<decltype(_Vec::allocator_type::_Minimum_allocation_alignment)>> =
+        _Vec::allocator_type::_Minimum_allocation_alignment >= 8;
+
+template <class _Ty, bool = _Has_minimum_allocation_alignment<_Ty>>
+class _NODISCARD _Asan_annotater {
+private:
+    using _Alloc      = typename _Ty::allocator_type;
+    using _Pointer    = typename _Ty::pointer;
+    using _Size_type  = typename _Ty::size_type;
+    using _Value_type = typename _Ty::value_type;
+
+public:
+    explicit constexpr _Asan_annotater(const _Ty& _Target) noexcept : _Target_vec(_Target), _Old_last(nullptr) {}
+
+    _CONSTEXPR20 _Asan_annotater(const _Ty& _Target, const _Size_type _New_size)
+        : _Target_vec(_Target), _Old_last(nullptr) {
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) {
+            return;
+        }
+#endif // _HAS_CXX20
+
+        const auto& _My_data     = _Target_vec._Mypair._Myval2;
+        const _Pointer& _Myfirst = _My_data._Myfirst;
+        const _Pointer& _Mylast  = _My_data._Mylast;
+        const _Pointer& _Myend   = _My_data._Myend;
+
+        if (_New_size > static_cast<_Size_type>(_Myend - _Myfirst)) {
+            return;
+        }
+
+        _Old_last = static_cast<const void*>(_Mylast);
+
+        if (_New_size > static_cast<_Size_type>(_Mylast - _Myfirst)) {
+            __sanitizer_annotate_contiguous_container(_Myfirst, _Myend, _Old_last, _Myfirst + _New_size);
+        }
+    }
+
+    _CONSTEXPR20 ~_Asan_annotater() {
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) {
+            return;
+        }
+#endif // _HAS_CXX20
+
+        const auto& _My_data     = _Target_vec._Mypair._Myval2;
+        const _Pointer& _Myfirst = _My_data._Myfirst;
+        const _Pointer& _Mylast  = _My_data._Mylast;
+        const _Pointer& _Myend   = _My_data._Myend;
+
+        if (!_Old_last) {
+            __sanitizer_annotate_contiguous_container(_Myfirst, _Myend, _Myend, _Mylast);
+        } else if (static_cast<const void*>(_Mylast) < _Old_last) {

```

Comment:
BUG: when the new size fits in the existing capacity, we increase ASAN's notion of size (in the constructor) before construction of the new element(s). If that construction throws, we fail to reduce ASAN's notion of size here in the destructor. I _think_ the fix is for this `else` clause to be performed unconditionally. (Applies to the other specialization.) (Indicates lack of test coverage.)

Reply:
Yup, that was a TODO